### DISCUSSION TITLE
Dumping out a producer config efficiently

### DISCUSSION THREAD

**Discussion by dc-oe** [edenhill/librdkafka#3498] [Q&A] [answered]
I'm looking to log the producer config after we create one. I grab the conf from the producer using rd_kafka_conf, which is specified as const. However, rd_kafka_conf_dump does not specify the const qualifier, causing me to duplicate the conf before dumping. Looking at the code, there is no reason why the const qualifier is not in the dump API.

Is this a bug, or is there a reason to not allow dumping a config from an instantiated producer?

Thanks
Dave

**Comment by edenhill** [comment, marked as answer]
That's a mistake for the dump() function signature, it should be const and it is treated as const internally. So you can safely cast your retrieved config to be un-const when calling this function.

Also have a look at `debug=conf` which prints the config on client initialization.

**Reply by dc-oe** [reply]
We did try the debug=conf property after I noticed it in the code, but we must be running an earlier version.

**Reply by dc-oe** [reply]
@edenhill  What I am looking for is to dump all properties and redact sensitive ones. This would allow devs to control.

const char **rd_kafka_conf_dump (const rd_kafka_conf_t *conf, size_t *cntp, rd_bool_t only_modified, rd_bool_t redact_sensitive) {
	return rd_kafka_anyconf_dump(_RK_GLOBAL, conf, cntp,
                                     only_modified,
                                     redact_sensitive);
}

Be willing to create a PR if it is something you would consider. It doesn't look like this is an option even in debug logging.

**Reply by edenhill** [reply]
That's pretty much what debug=conf does already. Any reason you can't use that?

**Reply by dc-oe** [reply]
It only dumps out changed properties. No option for dumping out all properties while redacting sensitive ones.

**Reply by edenhill** [reply]
Okay, in that case I suggest exposing rd_kafka_[topic_]conf_is_modified() and rd_kafka_[topic_]conf_is_sensitive() in the public API.
